File: Scala/src/main/scala/AlgorithmicHeights/MER.scala
package AlgorithmicHeights

object MER {
  import scala.annotation.tailrec

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  def mergeSortedLists(list1: List[Int], list2: List[Int]): List[Int] = {
    @tailrec
    def loop(acc: List[Int], xs: List[Int], ys: List[Int]): List[Int] = (xs, ys) match {
      case (Nil, Nil) => acc.reverse
      case (x :: xss, Nil) => loop(x :: acc, xss, ys)
      case (Nil, y :: yss) => loop(y :: acc, xs, yss)
      case (x :: xss, y :: yss) => if (x <= y) loop(x :: acc, xss, ys) else loop(y :: acc, xs, yss)
    }

    loop(Nil, list1, list2)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val _: Int = reader.next().toInt
    val list1: List[Int] = convertToIntList(reader.next())
    val _: Int = reader.next().toInt
    val list2: List[Int] = convertToIntList(reader.next())
    val result: List[Int] = mergeSortedLists(list1, list2)
    println(result.mkString(" "))
  }
}

File: OCaml/src/algorithmic_heights/mer.ml
let convert_to_intlist (line : string) : int list =
    List.map int_of_string Str.(line |> split (regexp " "))


let merge_sorted_lists (lst1 : int list) (lst2 : int list) : int list =
    let rec loop acc xs ys = match (xs, ys) with
        | ([], []) -> List.rev acc
        | (x :: xss, []) -> loop (x :: acc) xss ys
        | ([], y :: yss) -> loop (y :: acc) xs yss
        | (x :: xss, y :: yss) -> if x <= y then loop (x :: acc) xss ys else loop (y :: acc) xs yss
    in loop [] lst1 lst2


let () =
    ignore (read_int ());
    let xs = convert_to_intlist (read_line ()) in
    ignore (read_int ());
    let ys = convert_to_intlist (read_line ()) in
    let result = merge_sorted_lists xs ys in
    result |> List.map string_of_int |> String.concat " " |> print_endline
